pkg/controller/status.go

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+
+	ciliumv1alpha1 "github.com/gardener/gardener-extension-networking-cilium/pkg/apis/cilium/v1alpha1"
+	extensionscontroller "github.com/gardener/gardener-extensions/pkg/controller"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/retry"
+)
+
+func (a *actuator) updateProviderStatus(
+	ctx context.Context,
+	network *extensionsv1alpha1.Network,
+	config *ciliumv1alpha1.NetworkConfig,
+) error {
+	status, err := a.ComputeNetworkStatus(config)
+	if err != nil {
+		return err
+	}
+
+	return extensionscontroller.TryUpdateStatus(ctx, retry.DefaultBackoff, a.client, network, func() error {
```

Comment:
If you don't have retries then you can have two actors  (humand and controller) racing one another which can produce un-desired results, i.e., the controller will think it has updated the status but because the resourceVersion would be outdated  the one-time status update won't be enough and would produce a false-positive status update.

Reply:
The (desired) status should be set based on the context of the current reconciliation loop. 
It's the last step in the operation. In most cases you calculate whats the status which you expect the resource to have and the actual status. Update of the status should happen only if both differ.

If they differ and you try to update the status and that operation fails due to changed resource version of the resource, it might be a case that the specification of the resource might be changed. In this case it's the best to return error, requeue and go back to the top.